<a href="https://colab.research.google.com/github/natalalinda-cell/lenguajes_de_programacion/blob/main/ACTIVIDAD_VPN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Caso práctico: Analista de inversiones

Imagina que trabajas como **analista financiero** en una empresa de inversión.

Tu equipo debe decidir en cuál proyecto invertir.

La empresa solo puede elegir **un proyecto**.

Tu tarea será usar **Python para analizar la rentabilidad de cada proyecto** utilizando el **Valor Presente Neto (VPN)**.

## ¿Qué es el Valor Presente Neto (VPN)?

El **Valor Presente Neto (VPN)** permite determinar si un proyecto de inversión es rentable.

- Si **VPN > 0** → el proyecto genera valor.
- Si **VPN < 0** → el proyecto no es rentable.

Fórmula:

$$VPN = \\sum \
rac{Flujo_t}{(1+r)^t} - Inversión\ Inicial$$

## Proyectos de inversión

Usaremos **diccionarios en Python** para representar cada proyecto.

In [ ]:
proyecto_A = {
    'nombre': 'Proyecto Solar',
    'inversion': 8000,
    'flujos': [2500, 3000, 3500, 4000],
    'tasa': 0.10
}

proyecto_B = {
    'nombre': 'Proyecto Tecnológico',
    'inversion': 10000,
    'flujos': [3000, 3500, 4200, 4500],
    'tasa': 0.10
}

proyecto_C = {
    'nombre': 'Proyecto Logístico',
    'inversion': 7000,
    'flujos': [1800, 2600, 3200, 3800],
    'tasa': 0.10
}

## Calcular el VPN usando `for`

In [ ]:
vpn = -proyecto_A['inversion']

for t, flujo in enumerate(proyecto_A['flujos']):
    vpn += flujo / (1 + proyecto_A['tasa'])**(t+1)

print('VPN del proyecto A:', round(vpn, 3))

VPN del proyecto A: 2113.722


Decision con if para Proyecto A

In [ ]:
if vpn > 0:
    print('El proyecto es rentable')
else:
    print('El proyecto no es rentable')

El proyecto es rentable


## Ejercicio

Calcule el VPN del **Proyecto B** y **Proyecto C**.

In [ ]:
vpn=-proyecto_B['inversion']
for t, flujo in enumerate(proyecto_B['flujos']):
  vpn+=flujo/(1+proyecto_B['tasa'])**(t+1)
  print('VPN del proyecto B:', round(vpn, 2))

VPN del proyecto B: -7272.73
VPN del proyecto B: -4380.17
VPN del proyecto B: -1224.64
VPN del proyecto B: 1848.92


Decision con if para Proyecto B

In [ ]:
if vpn > 0:
    print('El proyecto es rentable')
else:
    print('El proyecto no es rentable')

El proyecto es rentable


In [ ]:
vpn = -proyecto_C['inversion']
for t, flujo in enumerate(proyecto_C['flujos']):
  vpn += flujo / (1 + proyecto_C['tasa'])**(t+1)

  print('VPN del proyecto C:', round(vpn, 2))


VPN del proyecto C: -5363.64
VPN del proyecto C: -3214.88
VPN del proyecto C: -810.67
VPN del proyecto C: 1784.78


Decision con if para Proyecto C

In [ ]:
if vpn > 0:
    print('El proyecto es rentable')
else:
    print('El proyecto no es rentable')

El proyecto es rentable


## Analizar todos los proyectos

In [ ]:
proyectos = [proyecto_A, proyecto_B, proyecto_C]

for proyecto in proyectos:
    vpn = -proyecto['inversion']

    for t, flujo in enumerate(proyecto['flujos']):
        vpn += flujo / (1 + proyecto['tasa'])**(t+1)

    print(proyecto['nombre'], 'VPN:', round(vpn, 2))

Proyecto Solar VPN: 2113.72
Proyecto Tecnológico VPN: 1848.92
Proyecto Logístico VPN: 1784.78


## Simulación de tasas con `while`

In [ ]:
tasa = 0.05

while tasa <= 0.20:
    vpn = -proyecto_A['inversion']

    for t, flujo in enumerate(proyecto_A['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

    print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: 3416.28
Tasa: 0.1 VPN: 2113.72
Tasa: 0.15000000000000002 VPN: 1030.66
Tasa: 0.2 VPN: 121.14


While para Proyecto B

In [ ]:
tasa = 0.05
while tasa <= 0.20:
    vpn = -proyecto_B['inversion']

    for t, flujo in enumerate(proyecto_B['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

        print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05


Tasa: 0.05 VPN: -7142.86
Tasa: 0.05 VPN: -3968.25
Tasa: 0.05 VPN: -340.14
Tasa: 0.05 VPN: 3362.03
Tasa: 0.1 VPN: -7272.73
Tasa: 0.1 VPN: -4380.17
Tasa: 0.1 VPN: -1224.64
Tasa: 0.1 VPN: 1848.92
Tasa: 0.15000000000000002 VPN: -7391.3
Tasa: 0.15000000000000002 VPN: -4744.8
Tasa: 0.15000000000000002 VPN: -1983.23
Tasa: 0.15000000000000002 VPN: 589.66
Tasa: 0.2 VPN: -7500.0
Tasa: 0.2 VPN: -5069.44
Tasa: 0.2 VPN: -2638.89
Tasa: 0.2 VPN: -468.75


While para Proyecto C

In [ ]:
tasa = 0.05
while tasa <= 0.20:
    vpn = -proyecto_C['inversion']
    for t, flujo in enumerate(proyecto_C['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

        print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: -5285.71
Tasa: 0.05 VPN: -2927.44
Tasa: 0.05 VPN: -163.16
Tasa: 0.05 VPN: 2963.11
Tasa: 0.1 VPN: -5363.64
Tasa: 0.1 VPN: -3214.88
Tasa: 0.1 VPN: -810.67
Tasa: 0.1 VPN: 1784.78
Tasa: 0.15000000000000002 VPN: -5434.78
Tasa: 0.15000000000000002 VPN: -3468.81
Tasa: 0.15000000000000002 VPN: -1364.76
Tasa: 0.15000000000000002 VPN: 807.91
Tasa: 0.2 VPN: -5500.0
Tasa: 0.2 VPN: -3694.44
Tasa: 0.2 VPN: -1842.59
Tasa: 0.2 VPN: -10.03


## Respuestas finales

1. Calcule el VPN de los tres proyectos.

Proyecto Solar VPN: 2113.72

Proyecto Tecnológico VPN: 1848.92

Proyecto Logístico VPN: 1784.78


# Determinar cual proyecto tiene el **VPN mas alto**

In [ ]:
for proyecto in proyectos:
    vpn = -proyecto['inversion']

    for t, flujo in enumerate(proyecto['flujos']):
        vpn += flujo / (1 + proyecto['tasa'])**(t+1)

    vpn_resultados[proyecto['nombre']] = round(vpn, 2)

proyecto_con_vpn_max = max(vpn_resultados, key=vpn_resultados.get)
vpn_maximo = vpn_resultados[proyecto_con_vpn_max]

print(f"El proyecto con el VPN más alto es '{proyecto_con_vpn_max}' con un VPN de: {vpn_maximo}")

El proyecto con el VPN más alto es 'Proyecto Solar' con un VPN de: 2113.72


# Justique la decision

El vpn positivo indica que todos los proyectos genrar valor para la empresa, pero como venmos en el cuadro de calculpo anterior el proyecto mas altos es (proyecto solar),  **2113.72**
lo convirte en el mas atactivo frente a
 (proyecto tecnologico) **1848.92**
 (proyecto logistico) **1784.78**
*proyecto solar tiene mayor capacidad de generar riqueza en terminos de VPN, siendo la decision mas solida desde la perpectiva financiera